backend/app/services/implementations/log_records_service.py

```diff
@@ -0,0 +1,157 @@
+
+        try:
+            new_log_record = LogRecords(**new_log_record)
+            db.session.add(new_log_record)
+            db.session.commit()
+            return log_record
+        except Exception as postgres_error:
+            raise postgres_error
+
+    def get_logs_list(self, logs):
+        try:
+            logs_list = []
+            # return as json object
+            for log in logs:
+                logs_list.append(
+                    {
+                        "log_id": log.log_id,
+                        "employee_id": log.employee_id,
+                        "resident_first_name": log.resident_first_name,
+                        "resident_last_name": log.resident_last_name,
+                        "datetime": str(log.datetime.astimezone(timezone("US/Eastern"))),
+                        "flagged": log.flagged,
+                        "attn_to": log.attn_to,
+                        "note": log.note,
+                        "tags": log.tags,
+                        "building": log.building
+
+                    }
+                )
+            return logs_list
+        except Exception as postgres_error:
+            raise postgres_error
+    
+    def filter_by_building(self, log_records, building):
+        return log_records.filter_by(building=building)
+    
+    def filter_by_employee_id(self, log_records, employee_id):
+        return log_records.filter_by(employee_id=employee_id)
+    
+    def filter_by_attn_to(self, log_records, attn_to):
+        return log_records.filter_by(attn_to=attn_to)
+    
+    def filter_by_date_range(self, log_records, date_range):
+        start_date = datetime.strptime(date_range[0], "%Y-%m-%d")
+        end_date = datetime.strptime(date_range[1], "%Y-%m-%d").replace(
+                hour=23, minute=59
+            )
+        return log_records.filter(
+            LogRecords.datetime >= start_date, LogRecords.datetime <= end_date
+        )
+
+    def filter_by_tags(self, log_records, tags):
+        for tag in tags:
+            log_records = log_records.filter(LogRecords.tags.any(tag))
+        return log_records
+    
+    def get_log_records(self, page_number,filters=None):
+        start_index = (page_number - 1) * RESULTS_PER_PAGE
+        end_index = start_index + RESULTS_PER_PAGE
+        if filters:
+            log_records = LogRecords.query.filter_by()
+            options = {
+                 "building": self.filter_by_building,
+                 "employee_id": self.filter_by_employee_id,
+                 "attn_to": self.filter_by_attn_to,
+                 "date_range": self. filter_by_date_range,
+                 "tags": self.filter_by_tags
+            }
+            for filter in filters:
+                log_records = options[filter](log_records, filters.get(filter))
+            num_results = log_records.count()
+            log_records = self.get_logs_list(log_records.order_by(LogRecords.datetime.desc())[start_index:end_index])
```

Comment:
Is this indexing safe here? Just curious about a case where we don't have enough logs to fill a page so we'll index out of bounds

Reply:
When I've been testing it on Postman its been fine when there isn't enough log records to fill a page, I'm pretty its just up to the `end_index` as a maximum